src/server/local-auth-helpers.js

```diff
@@ -0,0 +1,185 @@
+  if (nextUrl) validUuid(nextUrl, uuidMatch)
+
+  // Get salt and hash and verify user password
+  const pwFieldSplit = existingUser[0].auth0_id.split('|')
+  const hashed = {
+    salt: pwFieldSplit[1],
+    hash: pwFieldSplit[2]
+  }
+  return new Promise((resolve, reject) => {
+    AuthHasher.verify(
+      password, hashed,
+      (err, verified) => {
+        if (err) reject(err)
+        if (verified) {
+          resolve([null, existingUser[0]])
+        }
+        resolve([null, false, errorMessages.invalidCredentials])
+      }
+    )
+  })
+}
+
+const signup = async ({
+  lowerCaseEmail,
+  password,
+  existingUser,
+  nextUrl,
+  uuidMatch,
+  reqBody
+}) => {
+  // Verify UUID validity
+  if (nextUrl) validUuid(nextUrl, uuidMatch)
+
+  // Verify user doesn't already exist
+  if (existingUser.length > 0 && existingUser[0].email === lowerCaseEmail) {
+    return [null, false, errorMessages.emailTaken]
+  }
+
+  // Verify password and password confirm fields match
+  if (password !== reqBody.passwordConfirm) {
+    return [null, false, errorMessages.passwordsDontMatch]
+  }
+
+  // create the user
+  return new Promise((resolve, reject) => {
+    AuthHasher.hash(password, async function (err, hashed) {
+      if (err) reject(err)
+      // .salt and .hash
+      const passwordToSave = `localauth|${hashed.salt}|${hashed.hash}`
+      const user = await User.save({
+        email: lowerCaseEmail,
+        auth0_id: passwordToSave,
+        first_name: reqBody.firstName,
+        last_name: reqBody.lastName,
+        cell: reqBody.cell,
+        is_superadmin: false
+      })
+      resolve([null, user])
+    })
+  })
+}
+
+const reset = ({
+  password,
+  existingUser,
+  reqBody,
+  uuidMatch
+}) => {
+  if (existingUser.length === 0) {
+    return [null, false, errorMessages.invalidResetHash]
+  }
+
+  // Get user resetHash and date of hash creation
+  const pwFieldSplit = existingUser[0].auth0_id.split('|')
+  const [resetHash, datetime] = [pwFieldSplit[1], pwFieldSplit[2]]
+
+  // Verify hash was created in the last 15 mins
+  const isExpired = (Date.now() - datetime) / 1000 / 60 > 15
+  if (isExpired) {
+    return [null, false, errorMessages.resetHashExpired]
+  }
+
+  // Verify the UUID in request matches hash in DB
+  if (uuidMatch[0] !== resetHash) {
+    return [null, false, errorMessages.invalidResetHash]
+  }
+
+  // Verify passwords match
+  if (password !== reqBody.passwordConfirm) {
+    return [null, false, errorMessages.passwordsDontMatch]
+  }
+
+  // Save new user password to DB
+  return new Promise((resolve, reject) => {
+    AuthHasher.hash(password, async function (err, hashed) {
+      if (err) reject(err)
+      // .salt and .hash
+      const passwordToSave = `localauth|${hashed.salt}|${hashed.hash}`
+      const updatedUser = await User
+        .get(existingUser[0].id)
+        .update({ auth0_id: passwordToSave })
+        .run()
+      resolve([null, updatedUser])
+    })
+  })
+}
+
+const change = ({
```

Comment:
so this is the one endpoint that I think we should do through graphQL (or maybe I don't see it being done that way, but it is?  changing a password should require an authenticated person and we should also validate that the logged in user is the same as the one changing the password).

Otherwise, this endpoint is overly vulnerable

Reply:
OK, makes sense. I'll change this up.